google/cloud/sql/connector/instance_connection_manager.py

```diff
@@ -189,17 +190,24 @@ def __del__(self):
         """
         logger.debug("Entering deconstructor")
 
-        if self._current is not None:
-            logger.debug("Waiting for _current_instance_data to finish")
-            self._current.cancel()
-
-        if not self._client_session.closed:
-            logger.debug("Waiting for _client_session to close")
-            close_future = asyncio.run_coroutine_threadsafe(
-                self.__client_session.close(), loop=self._loop
-            )
-            close_future.result()
-
+        async def _deconstruct():
+            tasks = []
+            if not self._client_session.closed:
+                logger.debug("Waiting for _client_session to close")
+                tasks.append(self._client_session.close)
```

Comment:
You still need to call these: 
```suggestion
                tasks.append(self._client_session.close())
```

Remember, calling an async function returns the coro object, when then has to be awaited to get the actual results. 

asyncio.gather does the waiting part for you.

Reply:
I updated this to use `asyncio.gather(*[task() for task in tasks])`

Reply:
Any reason you decided to defer the function call? postponing it feels less clear to me